instrumentation/httpurlconnection/testing/src/androidTest/java/io/opentelemetry/instrumentation/library/httpurlconnection/InstrumentationTest.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.library.httpurlconnection;
+
+import static org.junit.Assert.assertEquals;
+
+import io.opentelemetry.android.test.common.OpenTelemetryTestUtils;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.sdk.testing.exporter.InMemorySpanExporter;
+import io.opentelemetry.sdk.trace.data.SpanData;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+public class InstrumentationTest {
+    private static final InMemorySpanExporter inMemorySpanExporter = InMemorySpanExporter.create();
+
+    @Before
+    public void setUp() {
+        OpenTelemetryTestUtils.setUpSpanExporter(inMemorySpanExporter);
+    }
+
+    @After
+    public void tearDown() {
+        inMemorySpanExporter.reset();
+    }
+
+    @Test
+    public void testHttpUrlConnectionGetRequest_ShouldBeTraced() {
+        HttpUrlConnectionTestUtil.executeGet("http://httpbin.org/get");
+        assertEquals(1, inMemorySpanExporter.getFinishedSpanItems().size());
+    }
+
+    @Test
+    public void testHttpUrlConnectionPostRequest_ShouldBeTraced() {
+        HttpUrlConnectionTestUtil.post("http://httpbin.org/post");
+        assertEquals(1, inMemorySpanExporter.getFinishedSpanItems().size());
+    }
+
+    @Test
+    public void
+            testHttpUrlConnectionGetRequest_WhenNoStreamFetchedAndNoDisconnectCalled_ShouldNotBeTraced() {
+        HttpUrlConnectionTestUtil.executeCustomGet("http://httpbin.org/get", false, false);
+        assertEquals(0, inMemorySpanExporter.getFinishedSpanItems().size());
+    }
+
+    @Test
+    public void
+            testHttpUrlConnectionGetRequest_WhenNoStreamFetchedButDisconnectCalled_ShouldBeTraced() {
+        HttpUrlConnectionTestUtil.executeCustomGet("http://httpbin.org/get", false, true);
+        assertEquals(1, inMemorySpanExporter.getFinishedSpanItems().size());
+    }
+
+    @Test
+    public void
+            testHttpUrlConnectionGetRequest_WhenNoStreamFetchedAndNoDisconnectCalledButHarvesterScheduled_ShouldBeTraced() {
+        HttpUrlConnectionTestUtil.executeCustomGet("http://httpbin.org/get", false, false);
+        ScheduledExecutorService harvester = scheduleHarvester();
+        try {
+            Thread.sleep(15000);
```

Comment:
I'd try to write the test in a way that we don't need to `sleep` that long, usually, this slows down the test pipeline quite fast

Reply:
➕ @surbhiia maybe you can somehow mock the time delay instead? We did something like that in the [PeriodicRunnableTest](https://github.com/open-telemetry/opentelemetry-android/blob/475f08f78c101d0b8d9e71e28134e933cec00a29/android-agent/src/test/java/io/opentelemetry/android/internal/services/periodicwork/PeriodicRunnableTest.kt#L28) class, not sure if the same approach would work though, but probably we can come up with some alternative anyway.

Reply:
One simpler approach that comes to mind is - when we make connection's inactivity timeout configurable (in next PR), we can set it 0s in this test case and run the harvester just once, right after, to end the trace, we can use ExecutorService and do a 2s awaitTimeout (mostly will finish before that) for this task to complete and do our assertion. Example:

`ExecutorService executorService = Executors.newSingleThreadExecutor();
  try {
        executorService.submit(HttpUrlInstrumentationConfig.getReportIdleConnectionRunnable());

        executorService.shutdown();

       if (executorService.awaitTermination(2, TimeUnit.SECONDS)) {
               Assert.assertEquals(1, inMemorySpanExporter.finishedSpanItems.size)
       } else {
               Assert.fail(
                    "Test could not be completed as tasks did not complete within the 2s timeout period.",
                )
        }
        } catch (InterruptedException e) {
         
        } finally {}
`

We can remove this test for now and add it in next PR (when we make timeout configurable), with this new approach. Does this look good?

Reply:
Another way to test this is to fake the clock this is using, but it requires a lot of plumbing will be beyond the scope of this change. 

I agree with @marandaneto - sleeps not only slow the test down, it can also introduce flakiness down the line because execution in CI is often unpredictable. So if you have to do it, make the value configurable so it won't take that long, and rely on `CountDownLatch` to verify conditions before proceeding rather than expecting certain things to be done after a certain amount of time.

Reply:
I've removed the harvester test from this PR. Will add it back in next PR when we make the inactivity timeout configurable, so we can set it to 0 and assert after 1 run of the harvester. I can wrap the runnable with a new runnable with CountDownLatch.